jwst/badpix_selfcal/badpix_selfcal.py

```diff
@@ -0,0 +1,93 @@
+from __future__ import annotations
+import numpy as np
+import jwst.datamodels as dm
+from jwst.outlier_detection.outlier_detection_ifu import medfilt
+from stdatamodels.jwst.datamodels.dqflags import pixel
+
+
+def badpix_selfcal(minimg: np.ndarray,
+                   flagfrac_lower: float = 0.001,
+                   flagfrac_upper: float = 0.001,
+                   kernel_size: int = 15,
+                   dispaxis=None) -> np.ndarray:
+    """
+    Flag residual artifacts as bad pixels in the DQ array of a JWST exposure
+
+    Parameters
+    ----------
+    minimg: np.ndarray
+        Selfcal data of shape (x, y), i.e., after some operation has
+        already been taken to combine multiple exposures,
+        typically a MIN operation.
+    flagfrac_lower: float
+        Fraction of pixels to flag on the low end
+    flagfrac_upper: float
+        Fraction of pixels to flag on the high end
+    kernel_size: int
+        Size of kernel for median filter
+    dispaxis: int
+        Dispersion axis, either 1 or 2. If None, a two-dimensional
+        median filter is applied.
+
+    Returns
+    -------
+    flagged_indices: np.ndarray
+        Indices of the flagged pixels,
+        shaped like output from np.where
+    """
+    if (flagfrac_lower < 0) or (flagfrac_upper < 0):
+        raise ValueError("Flagged fraction must be between 0 and 0.5")
+    if flagfrac_lower + flagfrac_upper >= 1:
+        raise ValueError("Total flagged fraction (lower plus upper) must be less than 1")
+
+    if dispaxis not in [1, 2, None]:
+        raise ValueError("dispaxis must be either 1 or 2, or None.")
+
+    # Determine outliers using median filter
+    if dispaxis is None:
+        kern_size = (kernel_size, kernel_size)
+    elif dispaxis == 2:
+        # check that this is the right way around!
+        kern_size = (kernel_size, 1)
+    elif dispaxis == 1:
+        kern_size = (1, kernel_size)
+
+    smoothed = medfilt(minimg, kern_size)
+    minimg_hpf = minimg - smoothed
+
+    # Flag outliers using percentile cutoff
+    flag_low, flag_high = np.nanpercentile(minimg_hpf, [flagfrac_lower * 100, (1 - flagfrac_upper) * 100])
+    bad = (minimg_hpf > flag_high) | (minimg_hpf < flag_low)
+    flagged_indices = np.where(bad)
+
+    return flagged_indices
+
+
+def apply_flags(input_model: dm.IFUImageModel, flagged_indices: np.ndarray) -> dm.IFUImageModel:
+    """
+    Apply the flagged indices to the input model. Sets the flagged pixels to NaN
+    and the DQ flag to DO_NOT_USE + OTHER_BAD_PIXEL
+
+    Parameters
+    ----------
+    input_model: dm.IFUImageModel
+        Input science data to be corrected
+    flagged_indices: np.ndarray
+        Indices of the flagged pixels,
+        shaped like output from np.where
+
+    Returns
+    -------
+    output_model: dm.IFUImageModel
+        Flagged data model
+    """
+
+    input_model.dq[flagged_indices] += pixel["DO_NOT_USE"] + pixel["OTHER_BAD_PIXEL"]
```

Comment:
I'm still looking for clarification about this @melanieclarke : my understanding of the DQ flag scheme was that the values were all factor-of-two different, such that encoding multiple flags was allowed by simply adding the flag values.  If this is the case, then why is the OR operator desired here?

Reply:
sorry, didn't see you responded already

Reply:
That's correct, but they are not factor of 2 different from themselves!  The or is needed in case the pixel already has one of the flags you are trying to set.  This is probably unlikely for DO_NOT_USE in this case, since they wouldn't be flagged as bad if they were already NaN, but it's very possible that a pixel might already have an OTHER_BAD_PIXEL flag (some existing NIRSpec darks use that value).  In that case, adding here would set the pixel to 2 * OTHER_BAD_PIXEL, which is a completely different flag.

Reply:
And I missed your reply while I was typing in a new one. :)